enhancements/ingress/header-case-adjustment.md

```diff
@@ -0,0 +1,335 @@
+
+#### As a cluster administrator, I have a legacy client that expects the HTTP `cache-control` header to be capitalized as `Cache-Control`.
+
+To satisfy this use-case, the cluster administrator can specify `Cache-Control`
+in `spec.httpHeaders.headerNameCaseAdjustments`:
+
+```
+oc -n openshift-ingress-operator patch ingresscontrollers/default --type=json --patch='[{"op":"add","path":"/spec/httpHeaders/headerNameCaseAdjustments/-","value":"Cache-Control"}]'
+```
+
+The IngressController then adjusts the `cache-control` response header as
+specified.
+
+### Implementation Details
+
+Implementing this enhancement required changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variable, `ROUTER_H1_CASE_ADJUST`, which specifies adjustments as a
+comma-delimited list of header names in the desired capitalizations.  The router
+parses the value of `ROUTER_H1_CASE_ADJUST` and translates it into
+`h1-case-adjust` settings.  If `ROUTER_H1_CASE_ADJUST` is nonempty, the router
+also specifies the `h1-case-adjust-bogus-server` setting.
+
+For example, if `ROUTER_H1_CASE_ADJUST` is set to `Host,X-Forwarded-For`, then
+the router specifies the following settings in the HAProxy configuration:
+
+```
+global
+  # ...
+  h1-case-adjust host Host
+  h1-case-adjust x-forwarded-for X-Forwarded-For
+  # ...
+
+defaults
+  # ...
+  option h1-case-adjust-bogus-client
+  # ...
+```
+
+When a Route has the `haproxy.router.openshift.io/h1-adjust-case=true`
+annotation, the router specifies the `h1-case-adjust-bogus-server` setting on
+the configuration for that Route:
+
+```
+backend ...
+  option h1-case-adjust-bogus-server
+```
+
+The ingress operator translates the value of the new
+`spec.httpHeaders.headerNameCaseAdjustments` API field into the format for the
+`ROUTER_H1_CASE_ADJUST` environment variable and sets this environment variable
+to that value in the router Deployment.
+
+### Risks and Mitigations
+
+OpenShift cannot infer which HTTP header names legacy clients and servers expect
+to be capitalized a certain way, so cluster administrators and application
+developers need to be aware of the issue and configure IngressControllers and
+Routes using this new configuration when updating to OpenShift 4.7.  A release
+note may mitigate this concern; for example:
+
+"OpenShift 4.7 updates to HAProxy 2.2, which down-cases HTTP header names by
+default (for example, `Host: xyz.com` is transformed to `host: xyz.com`).  If
+legacy applications are sensitive to the capitalization of HTTP header names,
+see the IngressController's new `spec.httpHeaders.headerNameCaseAdjustments` API
+field for a solution to accommodate these legacy applications until they can be
+fixed."
+
```

Comment:
One other risk that might be worth including (taken from the[ HAProxy Documentation](https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4-option%20h1-case-adjust-bogus-client)):

```
Applications which fail to properly process requests or responses may require
to temporarily use such workarounds to adjust header names sent to them for
the time it takes the application to be fixed. Please note that an
application which requires such workarounds might be vulnerable to content
smuggling attacks and must absolutely be fixed.
```

I'm not sure if this risk of content smuggling attacks is applicable to our Ingress Controller, so i'll defer to you on whether or not that blurb is worth mentioning (I don't entirely understand how header case adjustments could be a security risk).

Reply:
I can't figure out a connection between case adjustment and content smuggling.  I think the point is that an application that is broken in one way (requiring HTTP headers to be capitalized in a certain way) is likely to be broken in another way (not correctly handling chunked encoding).

Reply:
Fair enough. Thanks for double checking.